core/prisma/migrations/20241030175616_add_is_title_to_pubfield_pubtype_join_table/migration.sql

```diff
@@ -0,0 +1,97 @@
+
+
+-- Migrate so that all pub types will have some sort of title field
+BEGIN;
+
+-- First try to match on if there's an existing 'title' or 'name' field and set that field to isTitle=true
+WITH FirstMatchingPubField AS (
+    SELECT
+      "_PubFieldToPubType"."A" as "pubfield",
+      "_PubFieldToPubType"."B" as "pubtype",
+      pub_fields.slug,
+      ROW_NUMBER() OVER (PARTITION BY "_PubFieldToPubType"."B" ORDER BY pub_fields."createdAt" ASC) AS row_num
+    FROM
+      "_PubFieldToPubType"
+      JOIN pub_fields ON "_PubFieldToPubType"."A" = pub_fields.id 
+    WHERE
+      (pub_fields."schemaName" = 'String' OR pub_fields."schemaName" = 'RichText' OR pub_fields."schemaName" = 'Number')
+      AND (
+          pub_fields.name ILIKE '%title%'
+          OR pub_fields.name ILIKE '%name%'
+      )
+)
+UPDATE "_PubFieldToPubType"
+SET "isTitle" = true 
+FROM FirstMatchingPubField
+WHERE 
+  "_PubFieldToPubType"."A" = FirstMatchingPubField.pubfield
+  AND "_PubFieldToPubType"."B" = FirstMatchingPubField.pubtype 
+  AND FirstMatchingPubField.row_num = 1;
+
+-- If there's no match, just set the first field that is string, rich text, or number
+WITH FirstAvailablePubField AS (
+    SELECT
+      "_PubFieldToPubType"."A" as "pubfield",
+      "_PubFieldToPubType"."B" as "pubtype",
+      pub_fields."slug",
+      pub_types.name,
+      ROW_NUMBER() OVER (PARTITION BY "_PubFieldToPubType"."B" ORDER BY pub_fields."createdAt" ASC) AS row_num
+    FROM
+      "_PubFieldToPubType"
+      JOIN pub_fields ON "_PubFieldToPubType"."A" = pub_fields.id 
+      JOIN pub_types ON "_PubFieldToPubType"."B" = pub_types.id
+    WHERE
+      (pub_fields."schemaName" = 'String' OR pub_fields."schemaName" = 'RichText' OR pub_fields."schemaName" = 'Number')
+      AND NOT EXISTS (
+        SELECT 1
+        FROM "_PubFieldToPubType" AS "sub"
+        WHERE "sub"."B" = "_PubFieldToPubType"."B"
+        AND "sub"."isTitle" = true
+      )
+)
+UPDATE "_PubFieldToPubType"
+SET "isTitle" = true 
+FROM FirstAvailablePubField
+WHERE 
+  "_PubFieldToPubType"."A" = FirstAvailablePubField.pubfield
+  AND "_PubFieldToPubType"."B" = FirstAvailablePubField.pubtype 
+  AND FirstAvailablePubField.row_num = 1;
+
+INSERT INTO "_PubFieldToPubType" ("A", "B", "isTitle")
+SELECT
+  "newField"."id" AS "A",
+  pub_types.id AS "B",
+  true AS isTitle
+FROM
+  pub_types
+  LEFT JOIN LATERAL (
+    SELECT 
+      id, slug 
+    FROM pub_fields 
+    WHERE (pub_fields."schemaName" = 'String' OR pub_fields."schemaName" = 'RichText' OR pub_fields."schemaName" = 'Number')
+      AND (
+          pub_fields.name ILIKE '%title%'
+          OR pub_fields.name ILIKE '%name%'
+      )
+    AND pub_fields."communityId" = pub_types."communityId"
+    LIMIT 1
+  ) AS "newField" ON true
+WHERE
+  NOT EXISTS (
+    SELECT 1
+    FROM "_PubFieldToPubType"
+    WHERE "_PubFieldToPubType"."B" = pub_types.id
+    AND "_PubFieldToPubType"."isTitle" = true
+  )
+  AND "newField".id IS NOT NULL;
```

Comment:
what does this part intend to do?

from what i gather, it tries to add a new field to a pubtype that looks like `%title%` or `%name%` to pubtypes that do not have a `title` field yet, and sets it to have `isTitle: true` right? I see that that is described in the issue, but is that what we want to happen?

I also see that we are requiring a title field to be selected when creating a pubtype, so that's inline with this.

I am generally in favor of this approach: I think all pubtypes should have titles. But I'm not sure everyone agrees. I will put it in the slack as a discussion point!

Reply:
oops I meant to add a comment—this was meant to get at the last part of:

> One-time data migration that adds the designation to the first existing "title" or "name" fields; if none exists, it adds it to the first-created string, rich text, or number field; if none exists...I don't think there's one in the db, but could just create a "title" field in that case

I was hoping this wouldn't come up, but at least locally it comes up with the Arcadia seed—for instance, the pub type "PDF Download" only has a field "PDF" which is of type `FileUpload`, so this bit of the migration will detect that "PDF Download" doesn't have an `isTitle` field, and add `arcadia:title` to this pub type. It seems a little heavy handed, but we would need something like it if we do in fact want to require that all pub types have a title field

Reply:
removing these last two steps! migration becomes simpler :)